File: web-demo/src/engine/derive.ts
// Derived state. A vehicle has ONE status; Live Tracking, Dispatch and
// Maintenance all read it from here so the product can never contradict itself.

import type {
  Driver, MaintenanceSchedule, MaintenanceStatus, OrgSettings, SlaState, Task,
  Vehicle, VehicleStatus, DocStatus,
} from './types';

export type Freshness = 'LIVE' | 'STALE' | 'OFFLINE' | 'NEVER';

export function gpsFreshness(
  lastAt: number | undefined, now: number, liveS = 45, staleS = 300,
): { state: Freshness; ageS: number | null } {
  if (!lastAt) return { state: 'NEVER', ageS: null };
  const age = (now - lastAt) / 1000;
  if (age <= liveS) return { state: 'LIVE', ageS: age };
  if (age <= staleS) return { state: 'STALE', ageS: age };
  return { state: 'OFFLINE', ageS: age };
}

export function vehicleStatus(
  v: Vehicle, opts: { openWorkOrder?: boolean; now: number; settings: OrgSettings },
): VehicleStatus {
  const { openWorkOrder, now, settings } = opts;
  if (v.lifecycle === 'retired' || v.lifecycle === 'suspended' || v.lifecycle === 'planned') {
    return 'offline';
  }
  if (v.lifecycle === 'maintenance' || openWorkOrder) return 'maintenance';
  if (!v.lastPositionAt) return 'not_tracked';
  const { state } = gpsFreshness(v.lastPositionAt, now, settings.gpsLiveThresholdS,
                                settings.gpsStaleThresholdS);
  if (state === 'OFFLINE') return 'offline';
  if ((v.speedKph ?? 0) > 3) return 'moving';
  if (v.ignitionOn) return 'idle';
  return 'stopped';
}

const MAINT_ORDER: MaintenanceStatus[] =
  ['healthy', 'due_soon', 'due', 'in_workshop', 'overdue', 'critical'];

export function worstMaintenance(a: MaintenanceStatus, b: MaintenanceStatus): MaintenanceStatus {
  return MAINT_ORDER.indexOf(a) >= MAINT_ORDER.indexOf(b) ? a : b;
}

export function daysBetween(isoDate: string, now: number): number {
  const then = Date.parse(isoDate + 'T00:00:00Z');
  return Math.round((then - now) / 86400000);
}

export function maintenanceStatus(
  s: MaintenanceSchedule, odometerKm: number, now: number, settings: OrgSettings,
): MaintenanceStatus {
  let worst: MaintenanceStatus = 'healthy';
  const bump = (c: MaintenanceStatus) => {
    if (['healthy', 'due_soon', 'due', 'overdue', 'critical'].indexOf(c) >
        ['healthy', 'due_soon', 'due', 'overdue', 'critical'].indexOf(worst)) worst = c;
  };
  if (s.dueAtKm != null) {
    const left = s.dueAtKm - odometerKm;
    if (left <= -1000) bump('critical');
    else if (left < 0) bump('overdue');
    else if (left <= 50) bump('due');
    else if (left <= settings.maintenanceDueSoonKm) bump('due_soon');
  }
  if (s.dueAtDate) {
    const left = daysBetween(s.dueAtDate, now);
    if (left <= -30) bump('critical');
    else if (left < 0) bump('overdue');
    else if (left <= 1) bump('due');
    else if (left <= settings.maintenanceDueSoonDays) bump('due_soon');
  }
  return worst;
}

export function describeDue(
  s: MaintenanceSchedule, odometerKm: number, now: number,
): string {
  const bits: string[] = [];
  if (s.dueAtKm != null) {
    const left = s.dueAtKm - odometerKm;
    bits.push(`${Math.abs(Math.round(left)).toLocaleString()} km ${left < 0 ? 'overdue' : 'remaining'}`);
  }
  if (s.dueAtDate) {
    const left = daysBetween(s.dueAtDate, now);
    bits.push(`${Math.abs(left)} days ${left < 0 ? 'overdue' : 'remaining'}`);
  }
  return bits.join(' · ') || 'No threshold set';
}

export function maintenanceHealth(counts: {
  overdue: number; critical: number; due: number; dueSoon: number;
  openWorkOrders: number; repeatFailures: number;
}, weights: Record<string, number>): number {
  const w = {
    overdue: 25, critical: 30, due: 10, due_soon: 4,
    open_work_order: 6, repeat_failure: 15, ...weights,
  };
  const penalty =
    counts.overdue * w.overdue + counts.critical * w.critical + counts.due * w.due +
    counts.dueSoon * w.due_soon + counts.openWorkOrders * w.open_work_order +
    counts.repeatFailures * w.repeat_failure;
  return Math.max(0, Math.min(100, Math.round(100 - penalty)));
}

export function documentStatus(expiryIso: string, now: number, warnDays = 30): DocStatus {
  const days = daysBetween(expiryIso, now);
  if (days < 0) return 'expired';
  if (days <= warnDays) return 'expiring_soon';
  return 'valid';
}

export function slaState(t: Task, now: number, atRiskMinutes = 15): SlaState {
  if (!t.slaDueAt) return 'none';
  if (t.status === 'completed') return (t.completedAt ?? now) <= t.slaDueAt ? 'met' : 'breached';
  if (t.status === 'cancelled' || t.status === 'failed') return 'none';
  const reference = t.eta ?? now;
  if (reference > t.slaDueAt || now > t.slaDueAt) return 'breached';
  if (t.slaDueAt - reference <= atRiskMinutes * 60000) return 'at_risk';
  return 'on_track';
}

export function slaMinutesRemaining(t: Task, now: number): number | null {
  if (!t.slaDueAt) return null;
  return Math.round((t.slaDueAt - now) / 60000);
}

export function safetyScore(
  counts: Record<string, number>, distanceKm: number, weights: Record<string, number>,
): number {
  const w = {
    overspeed: 6, harsh_braking: 4, harsh_acceleration: 3.5, harsh_cornering: 3,
    route_deviation: 2, incident: 12, ...weights,
  } as Record<string, number>;
  const exposure = Math.max(distanceKm, 25) / 100;
  const penalty = Object.entries(counts)
    .reduce((s, [k, n]) => s + (w[k] ?? 1) * n, 0) / exposure;
  return Math.round(Math.max(0, Math.min(100, 100 - penalty)) * 10) / 10;
}

export function fatigueRisk(hours: number): 'low' | 'moderate' | 'elevated' | 'high' {
  if (hours >= 9) return 'high';
  if (hours >= 7) return 'elevated';
  if (hours >= 5) return 'moderate';
  return 'low';
}

export function bookValue(v: Vehicle, now: number): number {
  if (!v.purchaseValue || !v.purchaseDate) return 0;
  const years = Math.max(0, (now - Date.parse(v.purchaseDate + 'T00:00:00Z')) / (365.25 * 86400000));
  const life = Math.max(1, v.depreciationYears);
  const annual = (v.purchaseValue - v.residualValue) / life;
  return Math.round(Math.max(v.residualValue, v.purchaseValue - annual * years) * 100) / 100;
}

export function annualDepreciation(v: Vehicle): number {
  if (!v.purchaseValue) return 0;
  return Math.round(((v.purchaseValue - v.residualValue) / Math.max(1, v.depreciationYears)) * 100) / 100;
}

export function co2ForTrip(
  litres: number, kwh: number, fuel: string, s: OrgSettings,
): number {
  if (fuel === 'electric') return Math.round(kwh * s.co2PerKwh * 1000) / 1000;
  const f = fuel === 'petrol' ? s.co2PerLitrePetrol : s.co2PerLitreDiesel;
  return Math.round(litres * f * 1000) / 1000;
}

export const SEVERITY_RANK: Record<string, number> = {
  critical: 0, high: 1, medium: 2, low: 3,
};

export function driverFatigue(d: Driver) {
  return fatigueRisk(d.dutyHoursToday);
}

File: backend/app/services/derive.py
"""Derived state.

Spec 17: a vehicle has ONE source of truth. Live Tracking, Dispatch and
Maintenance must never store their own copy of "status" - they all call these
helpers, so the system can never contradict itself.
"""
from __future__ import annotations

from datetime import date, datetime, timedelta, timezone

from app.core.enums import (DocumentStatus, MaintenanceStatus, SlaState,
                            VehicleLifecycle, VehicleStatus)


def gps_freshness(last_position_at: datetime | None, *, live_s: int = 45,
                  stale_s: int = 300, now: datetime | None = None) -> tuple[str, float | None]:
    """Returns (LIVE|STALE|OFFLINE|NEVER, age_seconds). Never show old as live."""
    if last_position_at is None:
        return "NEVER", None
    now = now or datetime.now(timezone.utc)
    age = (now - last_position_at).total_seconds()
    if age <= live_s:
        return "LIVE", age
    if age <= stale_s:
        return "STALE", age
    return "OFFLINE", age


def vehicle_status(vehicle, *, has_open_work_order: bool = False,
                   live_s: int = 45, stale_s: int = 300,
                   idle_minutes: float = 3.0, now: datetime | None = None) -> str:
    """The single derived operational state of a vehicle."""
    if vehicle.lifecycle in (VehicleLifecycle.RETIRED, VehicleLifecycle.SUSPENDED,
                             VehicleLifecycle.PLANNED):
        return VehicleStatus.OFFLINE
    if vehicle.lifecycle == VehicleLifecycle.MAINTENANCE or has_open_work_order:
        return VehicleStatus.MAINTENANCE
    if vehicle.last_position_at is None:
        return VehicleStatus.NOT_TRACKED
    freshness, _age = gps_freshness(vehicle.last_position_at, live_s=live_s,
                                    stale_s=stale_s, now=now)
    if freshness == "OFFLINE":
        return VehicleStatus.OFFLINE
    speed = vehicle.last_speed_kph or 0.0
    if speed > 3.0:
        return VehicleStatus.MOVING
    if vehicle.ignition_on:
        now = now or datetime.now(timezone.utc)
        if vehicle.idle_since and (now - vehicle.idle_since).total_seconds() >= idle_minutes * 60:
            return VehicleStatus.IDLE
        return VehicleStatus.IDLE if speed <= 3.0 else VehicleStatus.MOVING
    return VehicleStatus.STOPPED


def maintenance_status(schedule, odometer_km: float, *, today: date | None = None,
                       due_soon_km: float = 500, due_soon_days: int = 14) -> str:
    """Mileage OR time rule - whichever triggers first (spec 7.5)."""
    today = today or date.today()
    km_left = schedule.km_remaining(odometer_km)
    days_left = schedule.days_remaining(today)

    worst = MaintenanceStatus.HEALTHY
    order = [MaintenanceStatus.HEALTHY, MaintenanceStatus.DUE_SOON,
             MaintenanceStatus.DUE, MaintenanceStatus.OVERDUE, MaintenanceStatus.CRITICAL]

    def bump(candidate):
        nonlocal worst
        if order.index(candidate) > order.index(worst):
            worst = candidate

    if km_left is not None:
        if km_left <= -1000:
            bump(MaintenanceStatus.CRITICAL)
        elif km_left < 0:
            bump(MaintenanceStatus.OVERDUE)
        elif km_left <= 50:
            bump(MaintenanceStatus.DUE)
        elif km_left <= due_soon_km:
            bump(MaintenanceStatus.DUE_SOON)
    if days_left is not None:
        if days_left <= -30:
            bump(MaintenanceStatus.CRITICAL)
        elif days_left < 0:
            bump(MaintenanceStatus.OVERDUE)
        elif days_left <= 1:
            bump(MaintenanceStatus.DUE)
        elif days_left <= due_soon_days:
            bump(MaintenanceStatus.DUE_SOON)
    return worst


def maintenance_health(*, overdue: int, critical: int, due: int, due_soon: int,
                       open_work_orders: int, repeat_failures: int,
                       weights: dict | None = None) -> int:
    """0-100 vehicle maintenance health. Weights are org-configurable (spec 7.2)."""
    w = {"overdue": 25, "critical": 30, "due": 10, "due_soon": 4,
         "open_work_order": 6, "repeat_failure": 15, **(weights or {})}
    penalty = (overdue * w["overdue"] + critical * w["critical"] + due * w["due"]
               + due_soon * w["due_soon"] + open_work_orders * w["open_work_order"]
               + repeat_failures * w["repeat_failure"])
    return max(0, min(100, round(100 - penalty)))


def document_status(expiry: date | None, *, today: date | None = None,
                    warn_days: int = 30) -> str:
    if expiry is None:
        return DocumentStatus.VALID
    today = today or date.today()
    days = (expiry - today).days
    if days < 0:
        return DocumentStatus.EXPIRED
    if days <= warn_days:
        return DocumentStatus.EXPIRING_SOON
    return DocumentStatus.VALID


def sla_state(task, *, now: datetime | None = None, at_risk_minutes: int = 15) -> str:
    """On track / at risk / breached, judged against ETA where one exists."""
    from app.core.enums import TaskStatus

    if task.sla_due_at is None:
        return SlaState.NONE
    now = now or datetime.now(timezone.utc)
    if task.status == TaskStatus.COMPLETED:
        return SlaState.MET if (task.completed_at or now) <= task.sla_due_at else SlaState.BREACHED
    if task.status in (TaskStatus.CANCELLED, TaskStatus.FAILED):
        return SlaState.NONE
    reference = task.eta or now
    if reference > task.sla_due_at or now > task.sla_due_at:
        return SlaState.BREACHED
    if (task.sla_due_at - reference) <= timedelta(minutes=at_risk_minutes):
        return SlaState.AT_RISK
    return SlaState.ON_TRACK


def sla_minutes_remaining(task, now: datetime | None = None) -> int | None:
    if task.sla_due_at is None:
        return None
    now = now or datetime.now(timezone.utc)
    return round((task.sla_due_at - now).total_seconds() / 60)


def safety_score(event_counts: dict[str, int], distance_km: float,
                 weights: dict | None = None) -> float:
    """0-100, normalised per 100 km so long shifts are not punished (spec 13.1)."""
    w = {"overspeed": 6.0, "harsh_braking": 4.0, "harsh_acceleration": 3.5,
         "harsh_cornering": 3.0, "route_deviation": 2.0, "incident": 12.0,
         **(weights or {})}
    exposure = max(distance_km, 25.0) / 100.0
    penalty = sum(w.get(k, 1.0) * v for k, v in event_counts.items()) / exposure
    return round(max(0.0, min(100.0, 100.0 - penalty)), 1)


def straight_line_depreciation(vehicle, *, today: date | None = None) -> float:
    """Current book value under straight-line depreciation (spec 20)."""
    if not vehicle.purchase_value or not vehicle.purchase_date:
        return 0.0
    today = today or date.today()
    years = max(0.0, (today - vehicle.purchase_date).days / 365.25)
    residual = vehicle.residual_value or 0.0
    life = max(1, vehicle.depreciation_years or 7)
    annual = (vehicle.purchase_value - residual) / life
    return round(max(residual, vehicle.purchase_value - annual * years), 2)


def annual_depreciation(vehicle) -> float:
    if not vehicle.purchase_value:
        return 0.0
    residual = vehicle.residual_value or 0.0
    life = max(1, vehicle.depreciation_years or 7)
    return round((vehicle.purchase_value - residual) / life, 2)


def co2_kg(*, litres: float = 0.0, kwh: float = 0.0, fuel_type: str = "diesel",
           settings_map: dict | None = None) -> float:
    s = settings_map or {}
    if fuel_type == "electric":
        return round(kwh * s.get("co2_kg_per_kwh", 0.061), 3)
    factor = (s.get("co2_kg_per_litre_petrol", 2.31) if fuel_type == "petrol"
              else s.get("co2_kg_per_litre_diesel", 2.68))
    return round(litres * factor, 3)
